compute_endpoint/globus_compute_endpoint/engines/high_throughput/executor.py

```diff
@@ -0,0 +1,940 @@
+            else:
+                log.error("No execution provider available")
+                r = None
+        return r
+
+    def scale_in(self, blocks):
+        """Scale in the number of active blocks by specified amount.
+
+        The scale in method here is very rude. It doesn't give the workers
+        the opportunity to finish current tasks or cleanup. This is tracked
+        in issue #530
+
+        Raises:
+             NotImplementedError
+        """
+        to_kill = self.blocks[:blocks]
+        if self.provider:
+            r = self.provider.cancel(to_kill)
+        return r
+
+    def _get_job_ids(self):
+        return list(self.block.values())
+
+    def status(self):
+        """Return status of all blocks."""
+
+        status = []
+        if self.provider:
+            status = self.provider.status(self.blocks)
+
+        return status
+
+    def shutdown(self, hub=True, targets="all", block=False):
+        """Shutdown the executor, including all workers and controllers.
+
+        This is not implemented.
+
+        Kwargs:
+            - hub (Bool): Whether the hub should be shutdown, Default:True,
+            - targets (list of ints| 'all'): List of block id's to kill, Default:'all'
+            - block (Bool): To block for confirmations or not
+
+        Raises:
+             NotImplementedError
+        """
+
+        log.info("Attempting HighThroughputExecutor shutdown")
+        # self.outgoing_q.close()
+        # self.incoming_q.close()
+        if self.queue_proc:
+            try:
+                self.queue_proc.terminate()
+            except AttributeError:
+                log.info("Executor interchange terminate skipped due to wrong context")
+            except Exception:
+                log.exception("Terminating the interchange failed")
+        log.info("Finished HighThroughputExecutor shutdown attempt")
+        return True
+
+    def _cancel(self, future):
+        """Attempt cancelling a task tracked by the future by requesting
+        cancellation from the interchange. Task cancellation is attempted
+        only if the future is cancellable i.e not already in a terminal
+        state. This relies on the executor not setting the task to a running
+        state, and the task only tracking pending, and completed states.
+
+        Parameters
+        ----------
+        future
+
+        Returns
+        -------
+        Bool
+        """
+
+        ret_value = future._cancel()
+        log.debug("Sending cancel of task_id:{future.task_id} to interchange")
+        if ret_value is True:
+            self.command_client.run(TaskCancel(future.task_id))
+            log.debug("Sent TaskCancel to interchange")
+        return ret_value
+
+
+CANCELLED = "CANCELLED"
+CANCELLED_AND_NOTIFIED = "CANCELLED_AND_NOTIFIED"
+FINISHED = "FINISHED"
+
+
+class HTEXFuture(concurrent.futures.Future):
+    def __init__(self, executor):
+        super().__init__()
+        self.executor = executor
```

Comment:
1. Can we add typing?  In this case, executor would be ... a subclass of an Engine?
2. Looks like our usage of this class immediately sets the `task_id`.  While this works, it'd be clearer to include it as a possibility to set in the constructor, and would enable type-checking elsewhere.  Please consider:
    ```python
    def __init__(self, executor, task_id: str | None = None):
        super().__init__()
        self.task_id = task_id
    ```
3. Given the potential to have a _lot_ of HTEXFuture's in play, please consider use of `__slots__`.  A small memory and performance improvement:
    ```python
    class HTEXFuture(concurrent.futures.Future):
        __slots__ = ("executor", "task_id")

        def __init__( ... ):
        ...
    ```

Reply:
Fixed in follow up commits.